std/uring.zig

```diff
@@ -0,0 +1,606 @@
+            .len = 0,
+            .union1 = linux.io_uring_sqe.union1{ .fsync_flags = fsync_flags },
+            .user_data = user_data,
+            .union2 = undefined,
+        };
+    }
+
+    pub fn queueNop(self: *Self, user_data: u64) !void {
+        const sqe = self.sq.getSQE() orelse return error.NoSpaceLeft;
+        sqe.* = linux.io_uring_sqe{
+            .opcode = linux.IORING_OP_NOP,
+            .flags = 0,
+            .ioprio = undefined,
+            .fd = undefined,
+            .off = undefined,
+            .addr = undefined,
+            .len = undefined,
+            .union1 = undefined,
+            .user_data = user_data,
+            .union2 = undefined,
+        };
+    }
+
+    /// Submit sqes to the kernel.
+    ///
+    /// Returns number of sqes submitted
+    pub fn submit(self: *Self) !u32 {
+        var submitted: u32 = undefined;
+
+        var ktail_next = self.sq.linuxSQ.getTail();
+        if (self.sq.linuxSQ.getHead() != ktail_next) {
+            submitted = @intCast(u32, self.sq.linuxSQ.array.len);
+        } else if (self.sq.sqe_head == self.sq.sqe_tail) {
+            // nothing to submit
+            return 0;
+        } else {
+            // Fill in sqes that we have queued up, adding them to the kernel ring
+            submitted = 0;
+            while (self.sq.sqe_head + submitted < self.sq.sqe_tail) : (submitted += 1) {
+                const sqe_index = self.sq.sqe_head & self.sq.linuxSQ.ring_mask;
+                self.sq.linuxSQ.array[(ktail_next +% submitted) & self.sq.linuxSQ.ring_mask] = sqe_index;
+            }
+
+            if (submitted > 0) {
+                // Write barrier ensures that the SQE stores are updated
+                // with the tail update. This is needed so that the kernel
+                // will never see a tail update without the preceeding sQE
+                // stores being done.
+                @fence(.SeqCst);
+
+                self.sq.sqe_head += submitted;
+                self.sq.linuxSQ.setTail(ktail_next +% submitted);
+
+                // The kernel has the matching read barrrier for reading the
+                // SQ tail.
+                @fence(.SeqCst);
+            }
+        }
+
+        // const res = linux.io_uring_enter(self.fd, submitted, 0, linux.IORING_ENTER_GETEVENTS, null);
+        const res = linux.io_uring_enter(self.fd, submitted, 0, 0, null);
+        const errno = linux.getErrno(res);
+        if (errno != 0) return os.unexpectedErrno(errno);
+
+        return @truncate(u32, res);
+    }
+
+    pub fn getCompletion(self: Self) !?linux.io_uring_cqe {
+        // TODO: return null if no operations are pending
+        while (true) {
+            if (self.cq.linuxCQ.peek(0)) |pcqe| {
+                // FIXME: Issue is here: it seems the read/dereference of pcqe happens before the kernel has written to it.
+                //os.nanosleep(0,0);
+                @fence(.SeqCst);
+                const result = pcqe.*;
```

Comment:
Read acquire fences must follow the read operation they are intended to protect.